### PR TITLE
Improve error messages from FromMultipart

### PR DESCRIPTION
**Pull request by maksbotan**
Hi,

I've tried using `servant-multipart` and observed that error messages from parsing are not very informative. Therefore I decided to introduce Lenient mode, analogous to Capture.

### CODE REVIEW

**Comment by maksbotan** [comment]
8.0 and 8.2 builds failed due to `<>` missing from Prelude (I've removed `Data.Monoid` import). How about we drop them and raise lower requirement on `base` to 4.8?

**Comment by fisx** [comment]
Thanks!  Good error messages are useful.  I will take a look at this soon.

> 8.0 and 8.2 builds failed due to `<>` missing from Prelude (I've removed `Data.Monoid` import). How about we drop them and raise lower requirement on `base` to 4.8?

I personally think supporting everything down to 8.0 is not too much to ask.  As long as we don't have any evidence that nobody cares, I wouldn't touch it.

Anyway that would be a different PR, no?

**Comment by maksbotan** [comment]
Okay, I've reverted that small change, now it should build on 8.0.

I have another thought: do we really need the "`fromMultipart failed:`" prefix in the error message? I suppose it should be OK to drop it and return the parser's message verbatim?

**Comment by fisx** [comment]
> Okay, I've reverted that small change, now it should build on 8.0.
> 
> I have another thought: do we really need the "`fromMultipart failed:`" prefix in the error message? I suppose it should be OK to drop it and return the parser's message verbatim?

The extra prefix could be helpful, depending on how unhelpful the parser is, which depends on the user of the library, so I'm not sure we should depend on it.

Would `"Could not decode multipart mime body: <parser error message>"` be an improvement?  I'd prefer something like this because it does not leak servant internals to the API.

File: src/Servant/Multipart.hs [hunk 1]
```diff
@@ -246,10 +258,10 @@ class FromMultipart tag a where
   --   in a list of textual inputs and another list for
   --   files, try to extract a value of type @a@. When
   --   extraction fails, servant errors out with status code 400.
-  fromMultipart :: MultipartData tag -> Maybe a
+  fromMultipart :: MultipartData tag -> Either String a
```
Comment (fisx) [inline src/Servant/Multipart.hs:261]: There are ways to make this change easier for the library users, but I'm not sure it should be.  Perhaps it's good to force everybody to take a look at this if they can't somehow produce a useful error here; if not, it's still straight-forward enough to do something like `maybe (Left "no error info, sorry") Right` to the maybe-parser.

Yes, I think it should be a breaking change.  @arianvp, @dmjio anybody: any opinions?

File: test/Test.hs [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+  assertBody "false" res
+
+testBlogPostRawHandler :: Session ()
+testBlogPostRawHandler = do
+  res <- srequest $ buildRequestWithHeaders POST "/blogPostRaw" correctBody multipartHeaders
+  assertStatus 200 res
+  assertBody "title body" res
+
+correctBody :: BSL.ByteString
+correctBody = mconcat $ intersperse "\n"
```
Comment (fisx) [inline test/Test.hs:111]: again, feel free to ignore: marginally nicer (or worse, depending on your taste :) -

```suggestion
correctBody = cs $ unlines
```

File: test/Test.hs [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+  , "--XX"
+  , "Content-Disposition: form-data; name=\"body\"; filename=\"body.md\""
+  , ""
+  , "Foo body"
+  , ""
+  , "--XX--"
+  ]
+
+partialBody :: BSL.ByteString
+partialBody = mconcat $ intersperse "\n"
```
Comment (fisx) [inline test/Test.hs:125]:
```suggestion
partialBody = cs $ unlines
```

**Comment by fisx** [comment]
> > Okay, I've reverted that small change, now it should build on 8.0.
> > I have another thought: do we really need the "`fromMultipart failed:`" prefix in the error message? I suppose it should be OK to drop it and return the parser's message verbatim?
> 
> The extra prefix could be helpful, depending on how unhelpful the parser is, which depends on the user of the library, so I'm not sure we should depend on it.
> 
> Would `"Could not decode multipart mime body: <parser error message>"` be an improvement? I'd prefer something like this because it does not leak servant internals to the API.

Since the PR is such an obvious improvement either way: if you feel strongly about this and nobody else objects, I'm ok with the change you suggest.

**Comment by maksbotan** [comment]
Hi,

I'm okay with extended message hiding `fromMultipart`. I'll try to commit it today.

Re breaking changes: I believe there is no need in transition mechanism, let users see this right away and be able to react. Anyway I suppose this will go into major version bump, like 0.12.0.

Re nitpicks: I'll also react to them in a commit today :)

Thanks for the review!

**Comment by maksbotan** [comment]
I've rebased the history to make it pretty and incorporated your ideas. Thanks!

Just in case: after this is merged, I'm going to make a PR with regenerated haskell-ci and another one with version bump.

**Comment by maksbotan** [comment]
Ping? :)

**Comment by arianvp** [comment]
i'll try to give this a quick look tomorrow. if I don't then Matthias just should merge this I think

**Comment by fisx** [comment]
Thanks again, and sorry for the delay!

Not sure about when we'll make a release.  Probably not before [zurihac](https://github.com/haskell-servant/servant/issues/1299).

**Comment by maksbotan** [comment]
Thanks!

**Pull request merged**